frontend/class-breadcrumbs.php

```diff
@@ -771,33 +773,78 @@ private function get_link_info_for_term( $term ) {
 	 */
 	private function get_link_info_for_ptarchive( $pt ) {
 		$link          = array();
-		$archive_title = '';
-
-		if ( WPSEO_Options::get( 'bctitle-ptarchive-' . $pt, '' ) !== '' ) {
-
-			$archive_title = WPSEO_Options::get( 'bctitle-ptarchive-' . $pt );
-		}
-		else {
-			$post_type_obj = get_post_type_object( $pt );
-			if ( is_object( $post_type_obj ) ) {
-				if ( isset( $post_type_obj->label ) && $post_type_obj->label !== '' ) {
-					$archive_title = $post_type_obj->label;
-				}
-				elseif ( isset( $post_type_obj->labels->menu_name ) && $post_type_obj->labels->menu_name !== '' ) {
-					$archive_title = $post_type_obj->labels->menu_name;
-				}
-				else {
-					$archive_title = $post_type_obj->name;
-				}
-			}
-		}
+		$archive_title = $this->get_archive_title( $pt );
 
 		$link['url']  = get_post_type_archive_link( $pt );
 		$link['text'] = $archive_title;
 
 		return $link;
 	}
 
+	/**
+	 * Determines whether or not the passed post type has a custom set breadcrumb.
+	 *
+	 * @param string $post_type The post type to check.
+	 *
+	 * @return bool True if a custom breadcrumb was set.
+	 */
+	private function has_post_type_archive_breadcrumb( $post_type ) {
+		return WPSEO_Options::get( 'bctitle-ptarchive-' . $post_type, '' ) !== '';
+	}
+
+	/**
+	 * Determines whether or not the passed post_type is a WooCommerce product and if it has a breadcrumb title set.
+	 *
+	 * @param string $post_type The post type to check.
+	 *
+	 * @return bool True if a WooCommerce breadcrumb was found. False otherwise.
+	 */
+	private function has_woocommerce_breadcrumb( $post_type ) {
+		if ( ! WPSEO_Utils::is_woocommerce_active() || $post_type !== 'product' ) {
+			return false;
+		}
+
+		$shop_page_id = wc_get_page_id( 'shop' );
+
+		if ( $shop_page_id === -1 ) {
+			return false;
+		}
+
+		return WPSEO_Meta::get_value( 'bctitle', $shop_page_id ) !== '';
+	}
+
+	/**
+	 * Determines the archive title based on the passed post type.
+	 *
+	 * @param string $post_type The post type to determine the title for.
+	 *
+	 * @return string The archive title.
+	 */
+	private function get_archive_title( $post_type ) {
+		if ( $this->has_woocommerce_breadcrumb( $post_type ) ) {
+			return WPSEO_Meta::get_value( 'bctitle',  wc_get_page_id( 'shop' ) );
```

Comment:
In `has_woocommerce_breadcrumb` you are getting the value also. I would choose for an approach where `WPSEO_Meta::get_value` will be called once. This prevents the `wc_get_page_id` being called in multiple functions as well.